src/axom/slic/interface/slic.hpp

```diff
@@ -23,7 +24,7 @@ namespace axom
 {
 namespace slic
 {
-struct debug
+struct AXOM_API debug
```

Comment:
This change resolved a link error at the cost of introducing a link warning.  I haven't yet figured out how to resolve the warning, but that might have to wait until next time.

Fixes a link error in msvc (but introduces a link warning)

Link error that this fixes:
> LINK : warning LNK4286: symbol '?checksAreErrors@debug@slic@axom@@2_NA (public: static bool axom::slic::debug::checksAreErrors)' defined in 'slic.obj' is imported by 'DataStore.obj' [<axom>\mint\tests\mint_mesh_coordinates_test.vcxproj]

Link warning that this introduces:
> inlet_Inlet.obj : error LNK2001: unresolved external symbol "public: static bool axom::slic::debug::checksAreErrors" (?checksAreErrors@debug@slic@axom@@2_NA) [<axom>\inlet\tests\inlet_Inlet_test.vcxproj]
<axom>\build-axom2\bin\Debug\inlet_Inlet_test.exe : fatal error LNK1120: 1 unresolved externals [<axom>\inlet\tests\inlet_Inlet_test.vcxproj]

These repeat in multiple places.

I think this is somehow related to object libraries and building in shared mode (dlls) ?

Reply:
this looks like a classic export not quite working as expected issue.  How does AXOM_API change between the compilation setup and then when it is included in the test?  You should be able to dig into this by looking at the defines for each compile.

Reply:
Here's the current logic:
https://github.com/LLNL/axom/blob/c4e24c7ffe2e44001db5e33c74fc583b8d9c4bae/src/axom/config.hpp.in#L117-L144

Reply:
It expands to this in the generated `config.hpp`:
```cpp
/*
 * define proper lib exports for various platforms
 */

#define AXOM_WINDOWS_DLL_EXPORTS TRUE
#define CMAKE_WINDOWS_EXPORT_ALL_SYMBOLS ON

#if defined(AXOM_EXPORTS) || defined(axom_EXPORTS)
/* define catch all def */
#define AXOM_EXPORTS_DEFINED 1
#endif

#if defined(_WIN32)

#if defined(AXOM_WINDOWS_DLL_EXPORTS)

#if defined(AXOM_EXPORTS_DEFINED)
#define AXOM_API __declspec(dllexport)
#else
#define AXOM_API __declspec(dllimport)
#endif  /* defined(AXOM_EXPORTS_DEFINED) */

#else   /* defined(AXOM_WINDOWS_DLL_EXPORTS) */

/* empty for static builds */
#define AXOM_API

#endif  /* defined(AXOM_WINDOWS_DLL_EXPORTS) */
```

Reply:
I'll see if I can fix it via CMake dllimport/export magic

Reply:
during compilation are we defining `AXOM_EXPORTS` or `axom_EXPORTS` ?. CMake will automatically define `axom_EXPORTS` for the library `axom`, but if we are doing tricky things like object libs -- we may have to pass the define ourselves.

Reply:
Example from ascent where we explicitly provide the define b/c many libs may shape the `API` exports 
https://github.com/Alpine-DAV/ascent/blob/52cb26a30662a0cec6aa1f04239d12c0757a859d/src/ascent/CMakeLists.txt#L296

Reply:
In conduit we make sure all the lib names trigger the right exports: 

https://github.com/LLNL/conduit/blob/80cf13780e666d343f767f4b02dba3d95f42cabb/src/libs/blueprint/conduit_blueprint_exports.h.in#L22

Reply:
ending today's episode about
"the more you know (tm)" about dll exports

Reply:
I think I found the problem: https://github.com/LLNL/axom/pull/398/files#r537837113

... and I strongly suspect that using CMake's `GenerateExportHeaders` will resolve it.
See: #410

Reply:
either way -- the define needs to be there only when compiling, or else there will be a problem.

Reply:
See also: https://github.com/LLNL/axom/commit/66f601a535b39baed6185398ba73cef71aac7220

Reply:
Ok -- looks like I tracked it down.

CMake is supposed to add a private compiler define of the form `<target>_EXPORTS` to each target, which would trigger the proper branch (`dllImport` vs. `dllExport`). However, it was not adding those. I suspect the issue is due to our usage of object libraries.

I fixed this by explicitly adding a private `axom_EXPORTS` compiler define to each axom component.

Reply:
yep -- that's the issue I was trying to articulate, sorry if I didn't do a good job and you where chasing this too long.  Glad you fixed it.

Reply:
Thanks. Unfortunately, I only truly understood the problem after working through it and compiling in verbose mode :(.